exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java

```diff
@@ -100,61 +102,65 @@ abstract void writeExemplar(
 
   abstract void writeEof(Writer writer) throws IOException;
 
-  final void write(Collection<MetricData> metrics, OutputStream output) throws IOException {
-    Map<InstrumentationScopeInfo, List<MetricData>> metricsByScope =
-        metrics.stream()
-            // Not supported in specification yet.
-            .filter(metric -> metric.getType() != MetricDataType.EXPONENTIAL_HISTOGRAM)
-            // PrometheusHttpServer#getAggregationTemporality specifies cumulative temporality for
-            // all instruments, but non-SDK MetricProducers may not conform. We drop delta
-            // temporality metrics to avoid the complexity of stateful transformation to cumulative.
-            .filter(metric -> !isDeltaTemporality(metric))
-            .filter(metric -> metricNameFilter.test(metricName(metric)))
-            .collect(
-                Collectors.groupingBy(
-                    MetricData::getInstrumentationScopeInfo,
-                    LinkedHashMap::new,
-                    Collectors.toList()));
+  final HashSet<String> write(Collection<MetricData> metrics, OutputStream output)
+      throws IOException {
+    HashSet<String> conflictMetricNames = new HashSet<>();
+    Map<String, List<MetricData>> metricsByName = new LinkedHashMap<>();
+    Set<InstrumentationScopeInfo> scopes = new LinkedHashSet<>();
+    // Iterate through metrics, filtering and grouping by headerName
+    for (MetricData metric : metrics) {
```

Comment:
A for loop is more appropriate because we now need to track the unique scopes we've seen and metric name conflicts, which we return to the caller for logging purposes.